[external API] alerts: the renamening

hawkw opened the pull request:
In conversation with @ahl, we have determined that the external API for webhooks added in #7277 should be changed to focus on "alerts" as the first-class user-facing concept, with "webhooks" as one delivery mechanism for alerts. This way, we can talk about alerts as an entity in the API that exist independently of webhooks that deliver alerts, and the same alert types can be shared with other alert delivery mechanisms if any are added in the future.

What we currently refer to as "webhook events" and "webhook event classes" are therefore renamed to "alerts" and "alert classes". The current concept of "webhook receivers" is generalized to an "alert receiver" resource, of which webhook receivers are (currently) the only subtype. This way, if we add other mechanisms of delivering alerts in the future (email, first-class Slack integration, etc), we can introduce new subtypes of alert receivers. I've restructured the API to have both `/v1/alert-receivers/...` and `/v1/webhook-receivers/...` routes, with operations common to all alert receivers (list, view, add/remove subscriptions, delete)  under the `alert-receivers` route, and operations related to webhook-specific configuration (add/remove secrets, probe, deliveries) under the `webhook-receivers` route. I've also changed the `AlertReceiver` view to have a "kind" enum that stores the subtype-specific configuration; currently, this will only ever be "webhook", but I thought it was worth doing this now to make future additions cause less breakage for API consumers.

This is, admittedly, a somewhat large diff, but fortunately, most of it is just renaming stuff and moving it around. Reviewers can focus more or less exclusively to the changes to the external API routes and models, and maybe the database migrations. Any mistakes while renaming and moving things around have already been caught by the Rust compiler. :)

david-crespo commented:
I think this is good. One thing that throws me off (but I could definitely get used it) is that you create a webhook receiver with `POST /v1/webhook-receivers` but then you list and view it with `/v1/alert-receivers`. I can see why it works that way, and maybe with two kinds of receiver, the structure would be more obvious. On the other hand it's rare for end users to be looking at a list like what's in `nexus_tags.txt`. The docs site sidebar is the closest thing, but the structure there is not as visible as I'd like, mostly because we're using plain english for the titles, which I don't think we can avoid.

<img width="265" alt="image" src="https://github.com/user-attachments/assets/9cdb05cd-e390-4e4c-96ac-75cce0e56ec8" />

smklein commented on nexus/db-model/src/alert_subscription.rs:
```diff
@@ -0,0 +1,354 @@
+    Clone, Debug, Queryable, Selectable, Insertable, Serialize, Deserialize,
+)]
+#[diesel(table_name = alert_subscription)]
+pub struct AlertRxSubscription {
```
Note for myself: this was moved from `nexus/db-model/src/webhook_rx.rs`

smklein commented on nexus/db-queries/src/db/datastore/saga.rs:
```diff
@@ -450,9 +450,9 @@ mod test {
 
     #[tokio::test]
-    async fn test_create_event_idempotent() {
+    async fn test_create_alert_idempotent() {
```
We're renaming this from "test_create_event" to "test_create_alert" but it's still calling "saga_create_event"?

I think this is not an alert; I think this is a "steno" event.

smklein commented on nexus/db-queries/src/db/datastore/saga.rs:
```diff
@@ -494,7 +494,7 @@ mod test {
     async fn test_update_state_idempotent() {
         // Test setup
-        let logctx = dev::test_setup_log("test_create_event_idempotent");
+        let logctx = dev::test_setup_log("test_create_alert_idempotent");
```
Same as above; I don't think this is an alert

smklein commented on nexus/external-api/src/lib.rs:
```diff
@@ -3542,49 +3542,116 @@ pub trait NexusExternalApi {
+            method = DELETE,
+            path = "/v1/alert-receivers/{receiver}",
+            tags = ["system/alerts"],
+        }]
```
Interesting choices by rustfmt here, huh

hawkw replied:
oh that might actually be my fault, i don't know if rustfmt will format inside attributes like this...

smklein commented on nexus/external-api/src/lib.rs:
```diff
@@ -3542,49 +3542,116 @@ pub trait NexusExternalApi {
+        receiver: Query<params::AlertReceiverSelector>,
+        state_filter: Query<params::AlertDeliveryStateFilter>,
+        pagination: Query<PaginatedByTimeAndId>,
+    ) -> Result<HttpResponseOk<ResultsPage<views::WebhookDelivery>>, HttpError>;
```
Is it weird that this is titled "list alert-deliveries", but it's returning "webhook delivery" objects?

smklein replied:
namely: Should this be an "enum-of-one" if we expect there will be non-webhook alerts in the future?

hawkw replied:
So, I wasn't really sure what to do about this endpoint, honestly. Since it lists deliveries by receiver ID, the type of the response entries will all be the same type of delivery based on the type of receiver. 

They could be an enum, but ideally, there would be a single enum containing the _list_ of deliveries, so the user only has to handle the potentially variable type a single time, and once you inspect the top level enum, all the entries are the same type. I thought about doing this, but it felt a bit awkward with dropshot's `ResultsPage` --- though we could return an enum of multiple types of `ResultsPage`s.

Alternatively, we could just make this endpoint specific to receiver types, so this one could be `webhook-deliveries` and we could just add separate delivery list endpoints if we were to add other receiver types in future. That might be the better move, since then there's no enum at all..

ahl replied:
> So, I wasn't really sure what to do about this endpoint, honestly. Since it lists deliveries by receiver ID, the type of the response entries will all be the same type of delivery based on the type of receiver.

Why would we expect these to differ by receiver type?

> They could be an enum, but ideally, there would be a single enum containing the _list_ of deliveries, so the user only has to handle the potentially variable type a single time, and once you inspect the top level enum, all the entries are the same type. I thought about doing this, but it felt a bit awkward with dropshot's `ResultsPage` --- though we could return an enum of multiple types of `ResultsPage`s.

A ResultsPage of enums would be great; an enum of ResultsPage would fuck up pagination magic.

> Alternatively, we could just make this endpoint specific to receiver types, so this one could be `webhook-deliveries` and we could just add separate delivery list endpoints if we were to add other receiver types in future. That might be the better move, since then there's no enum at all..

Again, since multiple receiver types is speculative at this point, let's choose something relatively consistently and re-evaluate when we have a second receiver.

hawkw replied:
> > So, I wasn't really sure what to do about this endpoint, honestly. Since it lists deliveries by receiver ID, the type of the response entries will all be the same type of delivery based on the type of receiver.
> 
> Why would we expect these to differ by receiver type?

I thought I had commented on this somewhere else, but I see that it was actually for probes (https://github.com/oxidecomputer/omicron/pull/8169#discussion_r2093525465) --- but, it's the same reason. The `WebhookDeliveryAttempt` model contains an HTTP status code, which is specific to an HTTP-based delivery method such as webhooks. I'd expect something like an email receiver to not have that, and instead have some other kind of success/error information. 

Thinking about it, though, we could probably share the main "delivery" model across all receiver types and just have the "delivery attempt" model be the part that differs. I guess it would probably make sense to wrap the list of delivery attempts in the delivery model in an enum more than the individual delivery attempt, since it won't change on retries.

> A ResultsPage of enums would be great; an enum of ResultsPage would fuck up pagination magic.

Yeah, that's what I figured. I'll push it into the individual entry type.

hawkw replied:
Changed this to be a `ResultsPage<AlertDelivery>` where only the _attempt_ model (which has the status code) is webhook specific in 8e8b6d91bad48ee0f0fae79d2379341e71034030.

smklein commented on nexus/tests/integration_tests/endpoints.rs:
```diff
@@ -1246,7 +1249,7 @@ pub static DEMO_WEBHOOK_SECRET_CREATE: LazyLock<params::WebhookSecretCreate> =
     });
 
-// pub static DEMO_WEBHOOK_SUBSCRIPTION: LazyLock<shared::WebhookSubscription> =
+// pub static DEMO_ALERT_SUBSCRIPTION: LazyLock<shared::WebhookSubscription> =
```
Nit: I know this is commented out code (why is this here?) but should this be a `shared::AlertScubscription` now?

smklein commented on nexus/types/src/external_api/shared.rs:
```diff
@@ -377,16 +377,16 @@ impl JsonSchema for SwitchLinkState {
             std::sync::LazyLock::new(|| {
-                regex::Regex::new(WebhookSubscription::PATTERN).expect(
+                regex::Regex::new(AlertSubscription::PATTERN).expect(
                     "WebhookSubscription validation regex should be valid",
```
```suggestion
                    "AlertSubscription validation regex should be valid",
```

smklein commented on nexus/types/src/external_api/shared.rs:
```diff
@@ -395,7 +395,7 @@ impl WebhookSubscription {
         } else {
             Err(anyhow::anyhow!(
                 "webhook subscription {s:?} does not match the pattern {}",
```
```suggestion
                "alert subscription {s:?} does not match the pattern {}",
```

smklein commented on nexus/types/src/external_api/shared.rs:
```diff
@@ -406,35 +406,35 @@ impl WebhookSubscription {
-impl JsonSchema for WebhookSubscription {
+impl JsonSchema for AlertSubscription {
     fn schema_name() -> String {
         "WebhookSubscription".to_string()
```
```suggestion
        "AlertSubscription".to_string()
```

smklein replied:
(same below, for description / title in metadata)

hawkw commented:
Oh, @smklein, one other thing: there are a couple of places where we now have tables that have a few columns with unfortunate names ("event_class" and "event_id" rather than "alert_class" and "alert_id") because [CRDB doesn't support renaming columns idempotently](https://github.com/oxidecomputer/omicron/tree/main/schema/crdb#213-renaming-columns). Do you think it's worth changing the migrations to drop those tables and create new ones with nicer names, instead of just renaming the table?

ahl commented on nexus/external-api/src/lib.rs:
```diff
@@ -3660,79 +3693,48 @@ pub trait NexusExternalApi {
         method = POST,
-        path = "/v1/webhooks/receivers/{receiver}/probe",
-        tags = ["system/webhooks"],
+        path = "/v1/webhook-receivers/{receiver}/probe",
```
did you consider nesting webhook-receivers under alerts?

ahl replied:
@david-crespo can you comment? It seems like "as flat as possible... but not flatter" might be the addendum.

david-crespo replied:
In this case I think the middle paragraph is the real constraint, rather than the aesthetic norm about flatness. We don't have anywhere else in the API where we support both `/v1/alerts/1f733f7b-b2eb-429c-8b3a-69203cb55cd7` and `/v1/alerts/receivers/...`. I think that's worse than what we currently have in this PR. Though overall the difficulty here I think is the same one I was pointing to in https://github.com/oxidecomputer/omicron/pull/8169#issuecomment-2885200350 — the structure doesn't feel quite natural, though I concede that's an initial impression and I am totally open to the idea that a) it's not yet worth the effort required to get it to feel intuitive, since we don't know how the feature will evolve, and b) initial impression is less important than whether it's usable after spending a little time wiht it.

hawkw replied:
There is also the option of double-nesting, i.e. having `/v1/alerts/alerts/1f733f7b-b2eb-429c-8b3a-69203cb55cd7` and `/v1/alerts/receivers/my-cool-receiver`, I suppose. But, the "alerts/alerts" _feels_ unpleasant.

hawkw replied:
yeah, i really didn't like it either...

hawkw commented on schema/crdb/alerts-renamening/up04.sql:
Ughhhh apparently this won't work as written, since you can't "ALTER TYPE IF EXISTS":
```
──── STDERR:             omicron-nexus::test_all integration_tests::schema::dbinit_equals_sum_of_all_up
log file: /tmp/test_all-fef1124d0eceaf7d-dbinit_equals_sum_of_all_up.3128075.0.log
note: configured to log to "/tmp/test_all-fef1124d0eceaf7d-dbinit_equals_sum_of_all_up.3128075.0.log"

thread 'integration_tests::schema::dbinit_equals_sum_of_all_up' panicked at nexus/tests/integration_tests/schema.rs:70:38:
Failed to execute update: Error { kind: Db, cause: Some(DbError { severity: "ERROR", parsed_severity: Some(Error), code: SqlState(E42601), message: "at or near \"exists\": syntax error", detail: Some("source SQL:\nALTER TYPE IF EXISTS omicron.public.webhook_event_class\n              ^"), hint: Some("try \\h ALTER TYPE"), position: None, where_: None, schema: None, table: None, column: None, datatype: None, constraint: None, file: Some("lexer.go"), line: Some(271), routine: Some("Error") }) }
```

I guess we'll have to drop the previous enums and add new ones, that's a bummer...

ahl commented on nexus/external-api/src/lib.rs:
```diff
@@ -3660,79 +3693,48 @@ pub trait NexusExternalApi {
+        path = "/v1/webhook-receivers/{receiver}/probe",
+        tags = ["system/alerts"],
     }]
     async fn webhook_receiver_probe(
```
is this particular to webhook receivers? In the CLI we had decided to call this subcommand `oxide alert receiver probe` i.e. assuming it would not be specific to webhooks but general for all receivers.

hawkw replied:
it returns a response model that is webhook-specific (it contains information about the status code of the HTTP response etc).

we could make that model be an enum with a single variant if we think it's preferable for this to be a route that applies to all receiver types. but, my thinking was that it's possible there might be some future receiver types that cannot be probed (can you synchronously probe an email?)

hawkw replied:
Changed this to not be webhook-specific in eae40e0716a7ea376708fb15c13efe648722d925.

ahl commented on nexus/external-api/src/lib.rs:
```diff
@@ -3542,104 +3542,137 @@ pub trait NexusExternalApi {
+        path = "/v1/alert-deliveries",
+        tags = ["system/alerts"],
+    }]
+    async fn alert_delivery_list(
```
we had decided to call this `alert receiver log` in the CLI. It is particular to a specific receiver, right?

hawkw replied:
yes, it's particular to a receiver, which is taken as a query param. i think `log` is a good name for the CLI, but in the API it felt like it should be a "list of delivery resources"...it could also be `/v1/alert-receivers/{name-or-id}/deliveries` or something?

ahl replied:
> /v1/alert-receivers/{name-or-id}/deliveries

That's what I was thinking.

hawkw replied:
yeah, we should be able to do that. i'll note that the previous route was `/v1/webhook-deliveries?receiver=...` and i think that may also have been at @david-crespo's urging IIRC.

david-crespo replied:
Yep, `/v1/alert-receivers/{name-or-id}/deliveries` sounds good to me. I see we already have `/v1/alert-receivers` and subscriptions hanging off of it.

hawkw replied:
Done in eae40e0716a7ea376708fb15c13efe648722d925!

ahl commented on nexus/external-api/src/lib.rs:
```diff
@@ -3542,104 +3542,137 @@ pub trait NexusExternalApi {
     }]
-    async fn webhook_receiver_update(
+    async fn alert_receiver_subscription_add(
         rqctx: RequestContext<Self::Context>,
-        path_params: Path<params::WebhookReceiverSelector>,
-        params: TypedBody<params::WebhookReceiverUpdate>,
-    ) -> Result<HttpResponseUpdatedNoContent, HttpError>;
+        path_params: Path<params::AlertReceiverSelector>,
+        params: TypedBody<params::AlertSubscriptionCreate>,
+    ) -> Result<HttpResponseCreated<views::AlertSubscriptionCreated>, HttpError>;
 
-    /// Delete webhook receiver
+    /// Remove alert receiver subscription
     #[endpoint {
         method = DELETE,
-        path = "/v1/webhooks/receivers/{receiver}",
-        tags = ["system/webhooks"],
+        path = "/v1/alert-receivers/{receiver}/subscriptions/{subscription}",
+        tags = ["system/alerts"],
     }]
```
you made what I thought was a good suggestion in the CLI to call these `oxide alert receiver subscribe` and `oxide alert receiver unsubscribe`. Do you now prefer subscription add/remove? I think I prefer the former, but my only strong preference is that CLI and API match in this regard.

hawkw replied:
i like the "subscribe"/"unsubscribe" naming for the CLI. per [this comment](https://github.com/oxidecomputer/omicron/pull/7277#discussion_r2061004371) from @david-crespo we prefer to use verbs like "add"/"remove" in API routes because they're consistent with other API operations, rather than descriptive verbs like "subscribe"/"unsubscribe". personally i think i would strongly prefer the more descriptive verbs in the CLI and don't have strong preferences about whether we should also use that in the API...perhaps it's more important for both to be consistent with each other than to use the more descriptive verbs, i dunno...

ahl replied:
Sounds fine; just wanted to make sure we're making the decision eyes open.

hawkw replied:
i definitely think there's value in using the same verbs in the CLI and the API, because then the CLI becomes a sort of teaching tool for the API: if you've done something manually, you know exactly where to look if you want to do it programmatically. but, i'm not totally sure how strongly we've weighed that against other concerns in the past. are there any cases where we've intentionally chosen to use different verbs (or nouns!) in the CLI, or are they always the same?

ahl commented on nexus/external-api/src/lib.rs:
```diff
@@ -3542,104 +3542,137 @@ pub trait NexusExternalApi {
-        path = "/v1/webhooks/receivers/{receiver}/subscriptions/{subscription}",
-        tags = ["system/webhooks"],
+        method = POST,
+        path = "/v1/webhook-receivers",
```
should `alert` be the top-level path?

hawkw replied:
see my discussion of nesting in https://github.com/oxidecomputer/omicron/pull/8169#discussion_r2093516103

ahl reviewed (commented):
took a pass comparing what we had determined in the CLI with the changes you're proposing to the API

ahl commented:
> @david-crespo re:
> 
> > One thing that throws me off (but I could definitely get used it) is that you create a webhook receiver with `POST /v1/webhook-receivers` but then you list and view it with `/v1/alert-receivers`. I can see why it works that way, and maybe with two kinds of receiver, the structure would be more obvious.

What about creation via `/v1/alert-receivers/webhook` POST?

> One thing I considered doing is _also_ having a `GET /v1/webhook-receivers` route to list/view webhook receivers only, in addition to the `GET` routes for `/v1/alert-receivers`.

My suggestion is that we do one or the other for now i.e. "list all receivers" (which happen to just be webhooks) or "list webhook receivers" (which happen to be all receivers). In the future, I can't imagine the utility for listing JUST the webhooks, but I have been known to lack imagination!

> That would return the webhook-specific models, and the view route would 404 if you requested a receiver ID/name that was a type other than webhook. I can see a couple advantages of this: it makes the API feel a bit more "complete", and it also provides you a way to get a webhook-receiver-specific model without having to handle the enum if you _know_ the receiver you want is a webhook receiver (and similarly, it gives you a way to list only webhook receivers). On the other hand, it means we have two separate routes that list/view the same entities, which could be confusing for users, and it requires us to maintain more endpoints. What do you think? Is it worth adding routes like that?

Given that a second flavor of webhook is speculation at this point, I would suggest just pick something and re-evaluate when we have something more concrete in the future.

hawkw commented:
> > @david-crespo re:
> > > One thing that throws me off (but I could definitely get used it) is that you create a webhook receiver with `POST /v1/webhook-receivers` but then you list and view it with `/v1/alert-receivers`. I can see why it works that way, and maybe with two kinds of receiver, the structure would be more obvious.
> 
> What about creation via `/v1/alert-receivers/webhook` POST?

Unfortunately we can't have a route like that, as receivers are looked up by name, and there's an ambiguity as to whether `/webhook` is interpreted as a name of a receiver to look up or a fixed path segment (as i discussed in https://github.com/oxidecomputer/omicron/pull/8169#discussion_r2093516103)

hawkw commented:
> I see "delivey" in several places--code and comment--I suggest a `git grep`.

whoops, thanks! fixed in 35a69f53a9926f0dde1566fa88ffa0dbfa94651a

> No additional comments with regard to the API since last time

I think I've mostly addressed your suggestions with the exception of nesting everything under `/alerts` --- how do you feel about the current routes in https://github.com/oxidecomputer/omicron/pull/8169/files#diff-6bf7d489f2f9a1890ff490732783605e46510de2ae69b36302d135aa5f9ebc68 ?
